swagger_spec_validator/validator20.py

```diff
@@ -21,5 +22,101 @@ def validate_spec_url(spec_url):
+    :returns: `None` in case of success, otherwise raises an exception.
+    :raises: :py:class:`swagger_spec_validator.SwaggerValidationError`
+    :raises: :py:class:`jsonschema.exceptions.ValidationError`
+    """
     validate_json(spec_json, 'schemas/v2.0/schema.json')
-    # TODO: rich schema validation
+
+    apis, definitions = spec_json['paths'], spec_json['definitions']
+
+    validate_apis(apis)
+
+    validate_definitions(definitions)
+
+
+def validate_apis(apis):
+    """Validates the semantic errors in `paths` of the Spec.
+
+    :param apis: dict of all the paths
+    :returns: `None` in case of success, otherwise raises an exception.
+    :raises: :py:class:`swagger_spec_validator.SwaggerValidationError`
+    :raises: :py:class:`jsonschema.exceptions.ValidationError`
+    """
+    for api_name in apis:
+        api_body = apis[api_name]
+        api_params = api_body.get('parameters', [])
+        validate_duplicate_param(api_params)
+        for oper_name in api_body:
+            oper_body = api_body[oper_name]
+            op_params = oper_body.get('parameters', [])
+            validate_duplicate_param(op_params)
+            all_params = list(set(get_path_param_names(api_params) +
+                                  get_path_param_names(op_params)))
+            validate_unresolvable_path_params(api_name, all_params)
+
+
+def validate_definitions(definitions):
+    """Validates the semantic errors in `definitions` of the Spec.
+
+    :param apis: dict of all the definitions
+    :returns: `None` in case of success, otherwise raises an exception.
+    :raises: :py:class:`swagger_spec_validator.SwaggerValidationError`
+    :raises: :py:class:`jsonschema.exceptions.ValidationError`
+    """
+    for def_name in definitions:
+        definition = definitions[def_name]
+        required = definition.get('required', [])
+        props = definition.get('properties', {}).keys()
+        extra_props = list(set(required) - set(props))
+        msg = "Required list has properties not defined"
+        if extra_props:
+            raise SwaggerValidationError("%s: %s" % (msg, extra_props))
+
+
+def get_path_param_names(params):
+    """Fetch all the names of the path parameters of an operation
+
+    :param params: list of all the params
+    :returns: list of the name of the path params
+    """
+    return [param['name']
+            for param in params
+            if param['in'] == 'path']
+
+
+def validate_duplicate_param(params):
+    """Validate no duplicate parameters are present.
+    Uniqueness is determined by the combination of 'name' and 'in'.
+
+    :param params: list of all the params
+    :returns: `None` in case of success, otherwise raises an exception.
+    :raises: :py:class:`swagger_spec_validator.SwaggerValidationError`
+    """
+    seen = set()
+    msg = "Duplicate param found with (name, in)"
+    for param in params:
+        param_id = (param['name'], param['in'])
```

Comment:
That kind of proves the point, ```(param['name'], param['in'])``` is valid to be use as the key in a hashtable. nbd